crates/block-producer/src/state_view/mod.rs

```diff
@@ -31,7 +34,7 @@ pub struct DefaultStateView<S> {
 
     /// The account ID of accounts being modified by transactions currently in the block production
     /// pipeline. We currently ensure that only 1 tx/block modifies any given account (issue: #186).
-    accounts_in_flight: Arc<RwLock<BTreeSet<AccountId>>>,
+    accounts_in_flight: Arc<RwLock<BTreeMap<AccountId, Digest>>>,
```

Comment:
nit: the doc comment here needs to be updated.

But also, I'm wondering if we should introduce a wrapper struct around the `BTreeMap`. Maybe something like `AccountStateTracker` (though, probably a better name is needed). For now, it could look something like this:

```rust
pub struct AccountStateTracker(BTreeMap<AccountId, Digest>);

impl AccountStateTracker {
    pub fn add(id: AccountId, init_state: Digest, final_state: Digest) -> Result<(), SomeError> {
        // if init_state is inconsistent with what is already being tracked, return an error
        todo!("implement")
    }

    pub fn remove(id: AccountId, final_state: Digest) -> Result<(), SomeError> {
        // removes the entry for id from the internal map, only if final_state matches the
        // corresponding value
        todo!("implement")
    }

    // could also have other convenience methods here - e.g., for checking if an account update
    // is in flight
}
```
We could also use a slightly more sophisticated approach internally and define the tracker as:
```rust
pub struct AccountStateTracker(BTreeMap<AccountId, Vec<Digest>>);
```
Here, the vector of digests would keep a chain of updates currently in-flight. Not strictly necessary, but I think it will make checking for inconsistencies easier.

Reply:
I'll implement something like this using a `VecDeque`. New transaction are appended and states are removed from the front once they complete.